db/go_level_db.go

```diff
@@ -169,102 +162,104 @@ func (mBatch *goLevelDBBatch) Write() {
 
 //----------------------------------------
 // Iterator
+// NOTE This is almost identical to db/c_level_db.Iterator
+// Before creating a third version, refactor.
+
+type goLevelDBIterator struct {
+	source    iterator.Iterator
+	start     []byte
+	end       []byte
+	isReverse bool
+	isInvalid bool
+}
+
+var _ Iterator = (*goLevelDBIterator)(nil)
 
-// https://godoc.org/github.com/syndtr/goleveldb/leveldb#DB.NewIterator
-// A nil Range.Start is treated as a key before all keys in the DB.
-// And a nil Range.Limit is treated as a key after all keys in the DB.
-func goLevelDBIterRange(start, end []byte) *util.Range {
-	// XXX: what if start == nil ?
-	if len(start) == 0 {
-		start = nil
+func newGoLevelDBIterator(source iterator.Iterator, start, end []byte, isReverse bool) *goLevelDBIterator {
+	if isReverse {
+		panic("not implemented yet") // XXX
 	}
-	return &util.Range{
-		Start: start,
-		Limit: end,
+	source.Seek(start)
```

Comment:
shouldn't we check for nil?

Reply:
I don't think so, otherwise the tests would have failed.  It did fail for cleveldb and I fixed it.